src/diffCheck/segmentation/DFSegmentation.cc

```diff
@@ -6,95 +6,227 @@
+
+        // iterate through the mesh faces given as function argument
+        for (std::shared_ptr<diffCheck::geometry::DFMesh> face : referenceMesh)
+        {
+            std::shared_ptr<geometry::DFPointCloud> correspondingSegment;
+            std::shared_ptr<open3d::geometry::TriangleMesh> o3dFace;
+            o3dFace = face->Cvt2O3DTriangleMesh();
+
+            // Getting the center of the mesh face
+            Eigen::Vector3d faceCenter;
+            open3d::geometry::OrientedBoundingBox orientedBoundingBox = o3dFace->GetMinimalOrientedBoundingBox();
+            faceCenter = orientedBoundingBox.GetCenter();
+
+            if (face->NormalsFace.size() == 0)
+            {
+                o3dFace->ComputeTriangleNormals();
+                face->NormalsFace.clear();
+                for (auto normal : o3dFace->triangle_normals_)
+                {
+                    face->NormalsFace.push_back(normal.cast<double>());
+                }
+            }
+            
+            // Getting the normal of the mesh face
+            Eigen::Vector3d faceNormal = face->NormalsFace[0];
+
+            // iterate through the segments to find the closest ones compared to the face center taking the normals into account
+            Eigen::Vector3d segmentCenter;
+            Eigen::Vector3d segmentNormal;
+            double faceDistance = std::numeric_limits<double>::max();
+            for (auto segment : clusters)
+            {
+                for (auto point : segment->Points)
+                {
+                    segmentCenter += point;
+                }
+                segmentCenter /= segment->GetNumPoints();
+
+                for (auto normal : segment->Normals)
+                {
+                    segmentNormal += normal;
+                }
+                segmentNormal.normalize();
+
+                double currentDistance = (faceCenter - segmentCenter).norm() / std::abs(segmentNormal.dot(faceNormal));
+                // if the distance is smaller than the previous one, update the distance and the corresponding segment
+                if (currentDistance < faceDistance)
+                {
+                    correspondingSegment = segment;
+                    faceDistance = currentDistance;
+                }
+            }
+            
+            // get the triangles of the face.
+            std::vector<Eigen::Vector3i> faceTriangles = o3dFace->triangles_;
+
+            for (Eigen::Vector3d point : correspondingSegment->Points)
+            {
+                bool pointInFace = false;
+                /*
+                To check if the point is in the face, we take into account all the triangles forming the face.
+                We calculate the area of each triangle, then check if the sum of the areas of the tree triangles 
+                formed by two of the points of the referencr triangle and our point is equal to the reference triangle area 
```

Comment:
just a thought here but in a future refinement having a ktree or some sort of structure for quick queries would highly improove this instead of looping through the cloud's points ?

Reply:
Yes I aggree.